File: elixir/nth-prime/nth_prime.exs
defmodule Prime do

  @doc """
  Generates the nth prime.
  """
  @spec nth(non_neg_integer) :: non_neg_integer
  def nth(count), do: nth(count, 2)

  def nth(count, _) when count < 1, do: raise ArgumentError
  def nth(1, p), do: p
  def nth(count, p), do: nth(count-1, next_prime(p))

  defp next_prime(2), do: 3
  defp next_prime(n) do
    Stream.iterate(n + 2, &(&1 + 2))
    |> Stream.drop_while(&(!prime?(&1)))
    |> Enum.take(1)
    |> hd
  end

  defp prime?(n) when is_integer(n) do
    ceiling = round(:math.sqrt(n))
    if n == ceiling * ceiling do
      false
    else
      prime_test(n, 2, ceiling, ceiling)
    end
  end

  defp prime_test(_, i, i, h), do: i == h
  defp prime_test(n, i, j, h) do
    if rem(n, i) == 0 do
      i == h
    else
      prime_test(n, i + 1, j, h)
    end
  end
end

File: go/nth-prime/nth_prime.go
package prime

import "math"

func Nth(n int) (int, bool) {
	if n < 1 {
		return 0, false
	}
	p := int64(2)
	for i := 1; i < n; i++ {
		p = nextPrime(p)
	}
	return int(p), true
}

func nextPrime(n int64) int64 {
	if n == 2 {
		return 3
	}
	for n = n + 2; !isPrime(n); n += 2 {
	}
	return n
}

func isPrime(n int64) bool {
	ceiling := round(math.Sqrt(float64(n)))
	if n == ceiling*ceiling {
		return false
	}
	return primeTest(n, 2, ceiling, ceiling)
}

func primeTest(n, i, j, h int64) bool {
	if i == j {
		return i == h
	}
	if n%i == 0 {
		return i == h
	}
	return primeTest(n, i+1, j, h)
}

// Is there a built-in round() somewhere? There's no math.Round().
func round(f float64) int64 {
	i64 := int64(f)
	if i64 == int64(f+0.5) {
		return i64
	}
	return i64 + 1
}
